부산대 Android_김주송_5주차 과제 (STEP2)

jooiss opened the pull request:
## 코드 작성 시, 어려웠던 점
- DataBinding을 layout에 하는 방법과 activity에서 설정하는 방법을 함께 사용하다보니 더 혼동이 왔던 것 같습니다. 
- LiveData를 적절한 부분에 사용한 게 맞는지 하는 의문이 들었습니다. (예를 들어 MapViewModel에서 첫 화면인지 확인하는 _hasIntentData를 넣었는데 이게 맞게 사용한 건지 살짝 고민이 됩니다.)
- ViewModel에 삽입해야 할 코드와 Activity에 삽입해야 할 코드들의 분리가 아직 명확히 되지 않아 보충 공부를 해야겠다 다짐하였습니다.

## 중점적으로 리뷰해주셨으면 하는 부분
- 패키지를 잘 분류했는지에 대해 확인받고 싶습니다.
- 지도 Error 상황에서 따로 Activity를 만들지 않고 레이아웃만 만들어서 변경하게 했는데 따로 Activity를 만드는 것이 나을지 여쭤보고 싶습니다.
- LiveData나 DataBinding으로 더 효율적으로 관리할 수 있는 제가 놓친 부분이 있는지 궁금합니다.
---
항상 정성스러운 코멘트 남겨주셔서 감사합니다. 많이 노력하겠습니다! 😁

MyStoryG committed:
Initial commit

MyStoryG committed:
Initial commit

jooiss committed:
docs: Update README.md

jooiss committed:
feat: Design search layout

jooiss committed:
feat: Add place database

- PlaceContract : 테이블 이름 및 열 이름 정의
- PlaceDatabaseHelper : 데이터베이스 생성
- PlaceDatabaseAccess : 데이터베이스에 데이터 저장 및 삭제 기능

jooiss committed:
fix: Add trigger to prevent storing duplicate data

jooiss committed:
feat: Design search recyclerView and place recyclerView

- place_item : place list item layout
- search_item : search list item layout

jooiss committed:
feat: Set the layout to be displayed

- PlaceDataModel : Place 데이터 클래스
- PlaceRecyclerViewAcitivity : Place 목록을 보여줄 recycler view 제어
- PlaceRecyclerViewAdapter : Place 목록 adapter

jooiss committed:
docs: Update README.md

- STEP2 구현 기능 목록 추가

jooiss committed:
feat: Add search function of place list

jooiss committed:
refactor: change place list item

jooiss committed:
refactor: resize search list item layout

jooiss committed:
feat: Add storing function of record list

MyStoryG committed:
Initial commit

jooiss committed:
fix: Resolve conflict

jooiss committed:
fix: Exclude kakao map dependency

jooiss committed:
docs: Update README.md

jooiss committed:
feat: Add database insert function

jooiss committed:
fix: Resolve screen-off feature

jooiss committed:
feat: Set API key value for kakao API integration

jooiss committed:
feat: Add data transfer object file

jooiss committed:
feat: Add retrofit interface

jooiss committed:
refactor: Combine data activity and place activity

jooiss committed:
feat: Add retrofit object

jooiss committed:
feat: Update network security configuration

jooiss committed:
feat: Add database insertion function

jooiss committed:
refactor: Extract string to file

jooiss committed:
refactor: Change the search item design

jooiss committed:
feat: Reset the data at first

jooiss committed:
feat: Change to category search

jooiss committed:
docs: Update README.md

jooiss committed:
feat: Add kakao native app key

jooiss committed:
feat: Add dependency about kakao map SDK

jooiss committed:
feat: Design map layout

jooiss committed:
feat: Add map view function using the kakao map API

jooiss committed:
fix: Resolve file redundancy

jooiss committed:
fix: set text to be written in one line

jooiss committed:
refactor: Change DTO file name

jooiss committed:
fix: Make API key resource not be saved

jooiss committed:
fix: Call MapView method in time for activity

jooiss committed:
fix: Change lifecycle function about kakao map API

jooiss committed:
fix: Change CategoryGroup finding method

jooiss committed:
refactor: Separate the Retrofit network-related function

jooiss committed:
fix: Resolve conflict

jooiss committed:
부산대 Android_김주송_4주차 과제 (STEP0) (#19)

* Initial commit

* docs: Update README.md

* feat: Design search layout

* feat: Add place database

- PlaceContract : 테이블 이름 및 열 이름 정의
- PlaceDatabaseHelper : 데이터베이스 생성
- PlaceDatabaseAccess : 데이터베이스에 데이터 저장 및 삭제 기능

* fix: Add trigger to prevent storing duplicate data

* feat: Design search recyclerView and place recyclerView

- place_item : place list item layout
- search_item : search list item layout

* feat: Set the layout to be displayed

- PlaceDataModel : Place 데이터 클래스
- PlaceRecyclerViewAcitivity : Place 목록을 보여줄 recycler view 제어
- PlaceRecyclerViewAdapter : Place 목록 adapter

* docs: Update README.md

- STEP2 구현 기능 목록 추가

* feat: Add search function of place list

* refactor: change place list item

* refactor: resize search list item layout

* feat: Add storing function of record list

* Initial commit

* fix: Exclude kakao map dependency

* docs: Update README.md

* feat: Add database insert function

* fix: Resolve screen-off feature

* feat: Set API key value for kakao API integration

* feat: Add data transfer object file

* feat: Add retrofit interface

* refactor: Combine data activity and place activity

* feat: Add retrofit object

* feat: Update network security configuration

* feat: Add database insertion function

* refactor: Extract string to file

* refactor: Change the search item design

* feat: Reset the data at first

* feat: Change to category search

* docs: Update README.md

* feat: Add kakao native app key

* feat: Add dependency about kakao map SDK

* feat: Design map layout

* feat: Add map view function using the kakao map API

* fix: Resolve file redundancy

* fix: set text to be written in one line

* refactor: Change DTO file name

* fix: Make API key resource not be saved

* fix: Call MapView method in time for activity

* fix: Change lifecycle function about kakao map API

* fix: Change CategoryGroup finding method

* refactor: Separate the Retrofit network-related function

---------

Co-authored-by: MyStoryG <bookwriterseo@gmail.com>

jooiss committed:
docs: Update README.md

jooiss committed:
refactor: Delete unnecessary files

jooiss committed:
feat: Change search method by keyword

jooiss committed:
feat: Display search result on selecting a saved search record

jooiss committed:
feat: Display modal when clicking on place list

jooiss committed:
feat: Show error message when onMapError

jooiss committed:
feat: Resume from last position on restart

jooiss committed:
test: Add UI test code for PlaceActivity

jooiss committed:
test: Add UI test code for MapActivity

jooiss committed:
test: Add unit test code for MapActivity

jooiss committed:
test: Add unit test code for PlaceActivity

jooiss committed:
Merge branch 'jooiss' into jooiss-step2

jooiss committed:
fix: Update UI test code for PlaceActivity

jooiss committed:
docs: Update README.md

jooiss committed:
부산대 Android_김주송_4주차 과제 (STEP1) (#71)

* Initial commit

* docs: Update README.md

* feat: Design search layout

* feat: Add place database

- PlaceContract : 테이블 이름 및 열 이름 정의
- PlaceDatabaseHelper : 데이터베이스 생성
- PlaceDatabaseAccess : 데이터베이스에 데이터 저장 및 삭제 기능

* fix: Add trigger to prevent storing duplicate data

* feat: Design search recyclerView and place recyclerView

- place_item : place list item layout
- search_item : search list item layout

* feat: Set the layout to be displayed

- PlaceDataModel : Place 데이터 클래스
- PlaceRecyclerViewAcitivity : Place 목록을 보여줄 recycler view 제어
- PlaceRecyclerViewAdapter : Place 목록 adapter

* docs: Update README.md

- STEP2 구현 기능 목록 추가

* feat: Add search function of place list

* refactor: change place list item

* refactor: resize search list item layout

* feat: Add storing function of record list

* Initial commit

* fix: Exclude kakao map dependency

* docs: Update README.md

* feat: Add database insert function

* fix: Resolve screen-off feature

* feat: Set API key value for kakao API integration

* feat: Add data transfer object file

* feat: Add retrofit interface

* refactor: Combine data activity and place activity

* feat: Add retrofit object

* feat: Update network security configuration

* feat: Add database insertion function

* refactor: Extract string to file

* refactor: Change the search item design

* feat: Reset the data at first

* feat: Change to category search

* docs: Update README.md

* feat: Add kakao native app key

* feat: Add dependency about kakao map SDK

* feat: Design map layout

* feat: Add map view function using the kakao map API

* fix: Resolve file redundancy

* fix: set text to be written in one line

* refactor: Change DTO file name

* fix: Make API key resource not be saved

* fix: Call MapView method in time for activity

* fix: Change lifecycle function about kakao map API

* fix: Change CategoryGroup finding method

* refactor: Separate the Retrofit network-related function

* docs: Update README.md

* refactor: Delete unnecessary files

* feat: Change search method by keyword

* feat: Display search result on selecting a saved search record

* feat: Display modal when clicking on place list

* feat: Show error message when onMapError

* feat: Resume from last position on restart

---------

Co-authored-by: MyStoryG <bookwriterseo@gmail.com>

LeeOhHyung committed:
Merge branch 'jooiss' into jooiss-step2

jooiss committed:
fix: Resolve merge conflict

jooiss committed:
refactor: Reorganize files according to data/domain/UI layer

jooiss committed:
chore: Rename application file name

jooiss committed:
refactor: Reorganize files according to clean architecture

jooiss committed:
refactor: Apply DataBinding in MapActivity and PlaceActivity

jooiss committed:
docs: Update README.md

jooiss committed:
feat: Add Place entity

jooiss committed:
feat: Add DAO interface

jooiss committed:
feat: Add Database

jooiss committed:
feat: Implement DBClient for managing Room database instance

jooiss committed:
refactor: Integrate Room database for data storage

jooiss committed:
test: Update test code for modified code

jooiss committed:
refactor: Add dependency injection for Retrofit

jooiss committed:
chore: Change Retrofit repository file name

jooiss committed:
refactor: Add dependency injection for Database

jooiss committed:
refactor: Add dependency injection for SharedPreferences

jooiss committed:
Merge branch 'jooiss' into jooiss-step1

jooiss committed:
fix: Resolve conflict

jooiss committed:
refactor: Separate Domain and Data layers for Place representation

jooiss committed:
feat: Handle duplicate entries to maintain data integrity

- Update schema to use [name, address] pair as unique identifier
- Implemente OnConflictStrategy for handling duplicates

jooiss committed:
refactor: Remove unnecessary code

jooiss committed:
feat: Add indices on 'name' and 'address' columns in PlaceEntity

jooiss committed:
refactor: Add dependency injection for latitude and longitude

jooiss committed:
docs: Update README.md

jooiss committed:
chore: Rename place layout file

jooiss committed:
refactor: Inject PlaceDao directly into PlaceRepository

jooiss committed:
refactor: Implement ViewModel and LiveData for PlaceActivity

- PlaceViewModel을 이용해 UI 관련 데이터와 로직 관리
- ViewModel의 LiveData를 관찰하여 UI를 업데이트하도록 리팩토링

jooiss committed:
refactor: Implement ViewModel and LiveData for MapActivity

- MapViewModel을 이용해 UI 관련 데이터와 로직 관리
- ViewModel의 LiveData를 관찰하여 UI를 업데이트하도록 리팩토링

jooiss committed:
refactor: Apply DataBinding for PlaceRecyclerViewAdapter

jooiss committed:
refactor: Apply DataBinding for SearchRecyclerViewAdapter

jooiss committed:
refactor: Add notifyDataSetChanged in RecyclerViewAdapter

jooiss committed:
refactor: Apply DataBinding for map error situation

jooiss committed:
refactor: Remove unnecessary code

jooiss committed:
fix: Resolve conflict

LeeOhHyung commented on app/src/main/java/campus/tech/kakao/map/ui/MapActivity.kt:
```diff
@@ -22,122 +21,42 @@ import com.kakao.vectormap.label.LabelOptions
+            viewModel.saveLocation()
+            viewModel.logSavedLocation()
+
+            viewModel.hasIntentData.observe(this@MapActivity, Observer { hasIntentData ->
+                if (hasIntentData) {
+                    displayMapLocation(kakaoMap)
+                    displayMarker()
+                    displayBottomSheet()
+                }
+            })
```
observer를 통해 데이터를 전달받고자 할때는 데이터를 생성해서 LiveData에 넘겨주기 이전에 Observer를 등록해두어야 합니다. 즉, 이 위치는 ReadyCallback 내부이므로 액티비티가 생성된 이후에 수행될걸로 보이네요. 이 보다는 onCreate 단계에서 하는것이 좀더 좋을것 같습니다. 

onMapReady 메소드가 여러번 호출되는지는 확실치 않네요. 만약 여러번 호출된다면, observe코드 또한 여러번 실행되고 등록된 옵저버도 다수가 될것이라 onCreate에서 한번만 수행되는것이 어떨까 합니다!

jooiss replied:
hasIntentData 삭제하면서 함께 삭제하였습니다!

LeeOhHyung commented on app/src/main/java/campus/tech/kakao/map/ui/MapViewModel.kt:
```diff
@@ -0,0 +1,73 @@
+    private val _hasIntentData = MutableLiveData<Boolean>()
+    val hasIntentData: LiveData<Boolean> get() = _hasIntentData
+
+    init {
+        _latitude.value = preferencesLatitude
+        _longitude.value = preferencesLongitude
+        _hasIntentData.value = false
+    }
+
+    fun updateLocationFromIntent(intent: Intent) {
+        _name.value = intent.getStringExtra("name")
+        _address.value = intent.getStringExtra("address")
+        _latitude.value = intent.getStringExtra("latitude") ?: preferencesLatitude
+        _longitude.value = intent.getStringExtra("longitude") ?: preferencesLongitude
+
+        _hasIntentData.value = !(_name.value.isNullOrEmpty() || _address.value.isNullOrEmpty() || _latitude.value.isNullOrEmpty() || _longitude.value.isNullOrEmpty())
+
+    }
```
>LiveData를 적절한 부분에 사용한 게 맞는지 하는 의문이 들었습니다. (예를 들어 MapViewModel에서 첫 화면인지 확인하는 _hasIntentData를 넣었는데 이게 맞게 사용한 건지 살짝 고민이 됩니다.)

아하, 이 코드의 의도를 알겠군요. 

MapActivity -> PlaceActivity -> MapActivity 로 돌아올때 name~longitude 값들을 전달받고자 하셨던걸로 보입니다. 이 경우에는 intent를 매번 확인하거나 hasIntentData 같은 변수를 선언하기 보다는 
Activity에서 제공해주는 `registerForActivityResult()` 메소드를 활용하면 더 좋을것 같습니다! ([참고](https://velog.io/@ho-taek/Android-registerForActivityResult%EB%9E%80))

Activity간의 전환에서 데이터를 주고받고, 결과가 올바르게 전달받았는지 확인할 수 있는 수단을 제공해줍니다.
이걸 사용하게 되면, PlaceActivity -> MapActivity로 갈때 `setResult(RESULT_OK, intent)` 하는 부분도 넣어주어야 겠네요!

jooiss replied:
registerForActivityResult로 변경하는 게 낫다는 걸 알면서도 많은 것을 수정해야 할 것 같아서 계속 저도 모르게 미뤘네요.. 바로 수정하였더니 더 효율적으로 진행하는 것 같아 너무 좋습니다. 감사합니다!

LeeOhHyung replied:
https://github.com/kakao-tech-campus-2nd-step2/android-map-refactoring/pull/72/commits/26e31387b91f4df1a62f821ea50122685bc32649 @jooiss 이 커밋에서 수정된것 확인했습니다! Activity <-> Activity 간에 데이터 전달 및 수신은 intent를 통해서 확인하는게 가장 일반적인 방법이긴합니다.

LeeOhHyung commented on app/src/main/java/campus/tech/kakao/map/ui/PlaceActivity.kt:
```diff
@@ -4,113 +4,50 @@ import android.content.Intent
-                }
-                addPlaceRecord(searchList, place)
-                controlSearchVisibility(searchList)
+        placeBinding.rvSearchList.adapter = searchAdapter
+        placeBinding.rvSearchList.layoutManager = LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false)
```
RecyclerView에 app:layoutManager="..." 속성으로 LayoutManager를 설정해줄 수 있습니다. 지금의 형태는 커스텀한 LayoutManager를 사용하는것이 아니기 때문에 XML속성으로도 사용가능해보이네요. 물론 Activity에서 하더라도 문제가 있는것은 아닙니다. UI 설정을 XML에서도 할 수 있는 방법이기때문에 참고하시면 좋을것 같습니다

jooiss replied:
몰랐던 방법입니다.. xml에 넣으니 더 코드가 깔끔해진 것 같습니다. 꿀팁 감사합니다!

LeeOhHyung replied:
@jooiss https://github.com/kakao-tech-campus-2nd-step2/android-map-refactoring/pull/72/commits/938bbd7a371d858c1ef22dcd9efc31be727d9415 이 커밋에서 수정된것 확인했습니다.

LeeOhHyung commented on app/src/main/java/campus/tech/kakao/map/ui/PlaceViewModel.kt:
```diff
@@ -0,0 +1,77 @@
+    private val networkRepository: NetworkRepository
+) : ViewModel() {
+
+    private val _placeList = MutableLiveData<List<Place>>()
+    val placeList: LiveData<List<Place>> get() = _placeList
+
+    private val _searchList = MutableLiveData<List<Place>>()
+    val searchList: LiveData<List<Place>> get() = _searchList
+
+    private val _isPlaceListVisible = MutableLiveData<Boolean>()
+    val isPlaceListVisible: LiveData<Boolean> get() = _isPlaceListVisible
+
+    private val _isSearchListVisible = MutableLiveData<Boolean>()
+    val isSearchListVisible: LiveData<Boolean> get() = _isSearchListVisible
```
>LiveData나 DataBinding으로 더 효율적으로 관리할 수 있는 제가 놓친 부분이 있는지 궁금합니다.

지금은 개수가 많이 없어서 필요한 데이터들을 하나하나 만들어도 되겠습니다. 하지만, 만약에 필요한 데이터의 종류가 10개, 20개 이상으로 많아진다면 관리하기 힘들어지는 시점이 옵니다. 저도 관리를 어떻게 하면 좋을지 고민이 많은데요.

data class를 사용해서 하나로 묶어주는 방법을 즐겨 사용합니다. 예를들면, 이렇게 되는것이죠! 

```kotlin
data class PlaceUiModel(
  val placeList: List<Place>,
  val searchList: List<Place>,
  ...
)

private val _uiModel = MutableLiveData<PlaceUiModel>()
val uiModel: LiveData<PlaceUiModel> get() = _uiModel
```

jooiss replied:
이런 방법이 있군요! 적용해서 코드 수정완료했습니다!

LeeOhHyung replied:
@jooiss https://github.com/kakao-tech-campus-2nd-step2/android-map-refactoring/pull/72/commits/8be1e46ccc78ac11d788a63cea4dabb0df45d908 이 커밋에서 수정하신것 확인했습니다. 훨씬 깔끔해진 느낌이군요

LeeOhHyung reviewed (changes requested):
@jooiss 이번 5주차도 고생많으셨습니다.
남긴 코멘트 참고하여 hasIntentData를 제거하는 방향으로 수정해보시면 좋을것 같습니다! 

그리고, Room을 통한 기기 저장소에 접근할때 뿐만 아니라 Retrofit을 통해 네트워크 통신을 할때도 코루틴을 사용해서 비동기통신을 할 수 있습니다. 여유가 되신다면 이것도 작업해보시면 더 좋을것 같습니다!

jooiss committed:
refactor: Replace startActivity with registerForActivityResult

- Retrieve data when the activity finishies
- Remove 'hasIntentData' LiveData for the presence of received data

jooiss committed:
refactor: Change LayoutManager configuration to XML

jooiss committed:
refactor: Use coroutines for asynchronous network calls with Retrofit

jooiss committed:
refactor: Consolidate Place-related LiveData into PlaceUiModel data class

jooiss commented:
코멘트 반영하여 코드 수정 완료하였습니다. 제가 잘못 적용한 부분이 있다면 편하게 말씀해주세요. 더 효율적인 코드 작성하게 도움주셔서 감사합니다!

- Retrofit 비동기 통신 적용
- intent로 전달받은 데이터가 있는지 확인하는 hasIntentData LiveData 삭제
- registerForActivityResult()를 이용하여 장소 목록 선택 시, 마커 표현 가능하도록 설정
- XML에서 LayoutManager 설정
- PlaceActivity UI와 관련된 LiveData를 PlaceUiModel 데이터 클래스로 관리

LeeOhHyung reviewed (approved):
@jooiss LGTM~ 
retrofit을 통한 api통신에도 suspend function으로 비동기통신하도록 수정하는것 확인했습니다. 빠르게 코멘트 반영해주셨네요. 고생많으셨습니다.
